cmd/snap-bootstrap/partition/disks.go

```diff
@@ -0,0 +1,499 @@
+	// is meant for testing/debugging.
+	String() string
+}
+
+type partition struct {
+	major    int
+	minor    int
+	label    string
+	partuuid string
+	path     string
+}
+
+type disk struct {
+	major      int
+	minor      int
+	partitions []*partition
+}
+
+func parseDeviceMajorMinor(s string) (int, int, error) {
+	errMsg := fmt.Errorf("invalid device number format: (expected <int>:<int>)")
+	devNums := strings.SplitN(s, ":", 2)
+	if len(devNums) != 2 {
+		return 0, 0, errMsg
+	}
+	maj, err := strconv.Atoi(devNums[0])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	min, err := strconv.Atoi(devNums[1])
+	if err != nil {
+		return 0, 0, errMsg
+	}
+	return maj, min, nil
+}
+
+var udevProperties = func(device string) (map[string]string, error) {
+	// now we have the partition for this mountpoint, we need to tie that back
+	// to a disk with a major minor, so query udev with the mount source path
+	// of the mountpoint for properties
+	cmd := exec.Command("udevadm", "info", "--name", device, "--query", "property")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, osutil.OutputErr(out, err)
+	}
+	r := bytes.NewBuffer(out)
+
+	return parseUdevProperties(r)
+}
+
+func parseUdevProperties(r io.Reader) (map[string]string, error) {
+	m := make(map[string]string)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		strs := strings.SplitN(scanner.Text(), "=", 2)
+		if len(strs) != 2 {
+			// bad udev output?
+			continue
+		}
+		m[strs[0]] = strs[1]
+	}
+
+	return m, scanner.Err()
+}
+
+// DiskFromMountPoint finds a matching Disk for the specified mount point.
+func DiskFromMountPoint(mountpoint string, opts *Options) (Disk, error) {
+	// call the unexported version that may be mocked by tests
+	return diskFromMountPoint(mountpoint, opts)
+}
+
+// the wrapper function just calls the implementation, but has the right
+// signature that returns a Disk, such that diskFromMountPointImplWrapper can
+// be assigned to diskFromMountPoint
+// the main reason we do this is so that we can call diskFromMountPointImpl from
+// other functions without type casting anything
+func diskFromMountPointImplWrapper(mountpoint string, opts *Options) (Disk, error) {
+	return diskFromMountPointImpl(mountpoint, opts)
+}
+
+func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
+	// first get the mount entry for the mountpoint
+	mounts, err := osutil.LoadMountInfo()
+	if err != nil {
+		return nil, err
+	}
+	found := false
+	d := &disk{}
+	mountpointPart := partition{}
+	for _, mount := range mounts {
+		if mount.MountDir == mountpoint {
+			mountpointPart.major = mount.DevMajor
+			mountpointPart.minor = mount.DevMinor
+			mountpointPart.path = mount.MountSource
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
+	}
+
+	majorMinor := fmt.Sprintf("%d:%d", mountpointPart.major, mountpointPart.minor)
+
+	if opts != nil && opts.IsDecryptedDevice {
```

Comment:
I haven't looked at the IsDecryptedDevice code, probably best is to split it out with its tests to a 2nd PR.